src/scripts/DataLayer.js

```diff
@@ -172,6 +172,35 @@ DataLayer.Manager.prototype._augment = function() {
     }
     return cloneDeep(that._state);
   };
+
+  /**
+   * Sets up a function that will be called whenever the specified event is triggered.
+   *
+   * @param {String} type A case-sensitive string representing the event type to listen for.
+   * @param {Function} listener A function that is called when the event of the specified type occurs.
+   * @param {Object} [options] Optional characteristics of the event listener.
+   * @returns {Number} The length of the data layer following push.
+   */
+  that._dataLayer.addEventListener = function(type, listener, options) {
+    return this.push({
```

Comment:
I wouldn't use `push()` as we should deprecate its support for event listeners. Instead I would use the internal implementations of it. 
The `options` argument should be supported.
Something along the lines:
```
  that._dataLayer.addEventListener = function(type, listenerFct, options) {
    const itemConfig = {
      on: type,
      handler: listenerFct,
      path: (options && options.path) ? options.path : null,
      scope: (options && options.scope) ? options.scope : null
    };
    const item = new DataLayer.Item(itemConfig);
    const listener = new DataLayer.Listener(item);
    for (const registeredItem of that._dataLayer) {
      that._listenerManager.triggerListener(listener, registeredItem);
    }
    that._listenerManager.register(listener);
  };

```

Reply:
I totally agree that if we stay with one API for add / remove event handler we should not use push internally to do that. Here is alternative approach which I took to avoid using push:
```  that._dataLayer.addEventListener = function(type, listener, options) {
    const eventItem = new DataLayer.Item({
      on: type,
      handler: listener,
      scope: (options && options.scope) || 'future'
    });

    that._processItem(eventItem);
  };``` What do you think about using internal `_proccessItem` method instead of duplicating its functionality?

Reply:
I think we should remove the following listener[On|Off] logic within _processItem() as we only want to support data, fctn and event and some of the logic is not needed anymore like `_getBefore(item)`.
````
    listenerOn: function(item) {
      const listener = new DataLayer.Listener(item);
      switch (listener.scope) {
        case DataLayer.constants.listenerScope.PAST:
          for (const registeredItem of _getBefore(item)) {
            that._listenerManager.triggerListener(listener, registeredItem);
          }
          break;
        case DataLayer.constants.listenerScope.FUTURE:
          that._listenerManager.register(listener);
          break;
        case DataLayer.constants.listenerScope.ALL:
          for (const registeredItem of _getBefore(item)) {
            that._listenerManager.triggerListener(listener, registeredItem);
          }
          that._listenerManager.register(listener);
      }
    },
    listenerOff: function(item) {
      that._listenerManager.unregister(new DataLayer.Listener(item));
    }
````

Reply:
I am not fully convinced of that approach. Firstly we still create an item in addEventListener:
```
item = new DataLayer.Item(itemConfig);
```
due to that, I think there should be a possibility to process each item type by `_proccess` method. I believe this is why there are multiple preProcessors inside of it. What do you think?

Reply:
Regarding scope: there have to be done quite some changes to the scope functionality (part of another ticket we have). Right now I am adjusting all tests to pass after removing push API for event handlers and make sure that scope remains as it is. I think we could follow up with scope after this PR or we can start working on scope changes meantime on this branch.

Reply:
you're right, my sample code above is lacking the scope logic. So, yes, we can do as you suggested:
`that._processItem(eventItem);`
Once we merge this PR, I will tackle the updates for the scope.